### PR TITLE
fix Gateway listener build bug

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
fixes: #11786

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 1]
```diff
@@ -57,19 +57,25 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env *model.Environme
...
-			// This is complicated. We have multiple servers. One of these servers could have passthrough HTTPS while
-			// others could be a simple/mutual TLS.
-			// The code as it is, is not capable of handling this mixed listener type and set up the proper SNI chains
-			// such that the passthrough ones go through a TCP proxy while others get terminated and go through http connection
-			// manager. Ideally, the merge gateway function should take care of this and intelligently create multiple
-			// groups of servers based on their TLS types as well. For now, we simply assume that if HTTPS,
-			// and the first server in the group is not a passthrough, then this is a HTTP connection manager.
-			if servers[0].Tls != nil && !model.IsPassThroughServer(servers[0]) {
-				protocol = model.ProtocolHTTP2
+		serversByProtocol := map[plugin.ListenerProtocol][]*networking.Server{}
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: There maybe two listener protocol for one port, like 443 TCP/HTTP

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 2]
```diff
@@ -108,30 +117,31 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env *model.Environme
...
+				}
 			}
-		}
 
-		// Filters are serialized one time into an opaque struct once we have the complete list.
-		if err := buildCompleteFilterChain(pluginParams, mutable, opts); err != nil {
-			errs = multierror.Append(errs, fmt.Errorf("gateway omitting listener %q due to: %v", mutable.Listener.Name, err.Error()))
-			continue
+			// Filters are serialized one time into an opaque struct once we have the complete list.
+			if err := buildCompleteFilterChain(pluginParams, mutable, opts); err != nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go]: I am not sure whether this is right for two different listener protocol types build filterChain twice

**Comment by hzxuzhonghu** [comment]
But I have tested as #11786, and it works fine.

**Comment by rshriram** [comment]
@hzxuzhonghu ptal at https://github.com/istio/istio/pull/12071 which does a more comprehensive fix.

**Pull request closed without merging**